amlib/dsp/stereo_convolution_mac.py

```diff
@@ -0,0 +1,439 @@
+            carryover2.eq(samples2_read_port.data[:self._bitwidth]),
+            carryover2_2.eq(carryover2),
+        ]
+
+        with m.FSM(reset="IDLE"):
+            with m.State("IDLE"):
+                # store new sample for left channel
+                with m.If(self.signal_in.valid & self.signal_in.first & ~set1):
+                    sample1_value = Cat(
+                        samples1_read_port.data[:-self._bitwidth],
+                        self.signal_in.payload.as_signed()
+                    )
+
+                    m.d.sync += [
+                        samples1_write_port.data.eq(sample1_value),
+                        samples1_write_port.addr.eq(0),
+                        samples1_write_port.en.eq(1),
+                        set1.eq(1),
+                    ]
+
+                # store new sample for right channel
+                with m.Elif(self.signal_in.valid & self.signal_in.last & ~set2):
+                    sample2_value = Cat(
+                        samples2_read_port.data[:-self._bitwidth],
+                        self.signal_in.payload.as_signed()
+                    )
+                    m.d.sync += [
+                        samples2_write_port.data.eq(sample2_value),
+                        samples2_write_port.addr.eq(0),
+                        samples2_write_port.en.eq(1),
+                        set2.eq(1),
+                    ]
+
+                # prepare MAC calculations
+                with m.If(set1 & set2):
+                    for i in range(self._slices * 2):
+                        m.d.sync += [
+                            ix.eq(0),
+                            madd_values[i].eq(0),
+                            previous_sample1.eq(0),
+                            previous_sample2.eq(0),
+                            current_sample1.eq(0),
+                            current_sample2.eq(0),
+                            carryover1.eq(0),
+                            carryover1_2.eq(0),
+                            carryover2.eq(0),
+                            carryover2_2.eq(0),
+                        ]
+
+                    m.next = "MAC"
+                with m.Else():
+                    m.d.comb += self.signal_in.ready.eq(1)
+            with m.State("MAC"):
+                # do the actual MAC calculation
+                with m.If(ix <= self._samples_per_slice):
+                    with m.If(ix > 0):
+                        for i in range(self._slices):
+                            left_sample = samples1_read_port.data[i*self._bitwidth:(i + 1) * self._bitwidth].as_signed()
+                            right_sample = samples2_read_port.data[i*self._bitwidth:(i + 1) * self._bitwidth].as_signed()
+                            main_tap = taps1_read_port.data[i*self._bitwidth:(i + 1) * self._bitwidth].as_signed()
+                            bleed_tap = taps2_read_port.data[i*self._bitwidth:(i + 1) * self._bitwidth].as_signed()
+
+                            #debugging
```

Comment:
Could you please make that code conditional to a debug flag that can be set by the user?

Reply:
Adjusted